shared/domains/chelonia/internals.js

```diff
@@ -681,13 +685,52 @@ export default (sbp('sbp/selectors/register', {
     }
     if (!state._vm) config.reactiveSet(state, '_vm', Object.create(null))
     const opFns: { [GIOpType]: (any) => void | Promise<void> } = {
+      /*
+        There are two types of "errors" that we need to consider:
+        1. "Ignoring" errors
+        2. "Failure" errors
+        Example: OP_KEY_ADD
+        1. IGNORING: an error is thrown because we wanted to add a key but the key we wanted to add is already there. This is not a hard error, it's an ignoring error. We don't care that the operation failed in this case because the intent was accomplished.
+        2. FAILURE: an error is thrown because we wanted to add a key that doesn't exist.
+        Example: OP_ACTION_ENCRYPTED
+        1. IGNORING: An error is thrown because we don't have the key to decrypt the action. We ignore it.
+        2. FAILURE: An error is thrown by the process function during processing.
+        Handling these in OP_ATOMIC
+        • ALL errors of class "IGNORING" should be ignored. They should not impact our ability to process the rest of the operations in the OP_ATOMIC. No matter how many of these are thrown, it doesn't affect the rest of the operations.
+        • ANY error of class "FAILURE" will call the rest of the operations to fail and the state to be reverted to prior to the OP_ATOMIC. No side-effects should be run. Because an intention failed.
+      */
       async [GIMessage.OP_ATOMIC] (v: GIOpAtomic) {
-        for (const u of v) {
-          if (u[0] === GIMessage.OP_ATOMIC) throw new Error('Cannot nest OP_ATOMIC')
-          if (!validateKeyPermissions(config, state, signingKeyId, u[0], u[1], direction)) {
-            throw new Error('Inside OP_ATOMIC: no matching signing key was defined')
+        for (let i = 0; i < v.length; i++) {
+          const u = v[i]
+          try {
+            if (u[0] === GIMessage.OP_ATOMIC) throw new Error('Cannot nest OP_ATOMIC')
+            if (!validateKeyPermissions(config, state, signingKeyId, u[0], u[1], direction)) {
+              throw new Error('Inside OP_ATOMIC: no matching signing key was defined')
+            }
+            await opFns[u[0]](u[1])
+          } catch (e) {
+            if (e) {
+              if (e.name === 'ChelErrorDecryptionKeyNotFound') {
+                console.warn(`[chelonia] [OP_ATOMIC] WARN '${e.name}' in processMutation for ${message.description()}: ${e.message}`, e, message.serialize())
+                const missingDecryptionKeyIds = missingDecryptionKeyIdsMap.get(message)
+                if (missingDecryptionKeyIds) {
+                  missingDecryptionKeyIds.add(e.cause)
```

Comment:
Are we guaranteed `e.cause` will be defined here?

Reply:
It's defined for that error, unless someone generates an error by the same name manually.